### PR TITLE
Implement HEAD requests support for artifacts [in progress]

### PR DESCRIPTION
**Pull request by sanket407**
Issue #372: Added Support for HEAD requests  for Last-Updated header.

TO-DO: 
* Add `getPath(String, String, String)` to `AbstractRepositoryProvider` class by resolving its dependency for RepositoryPath class
* Clean up `MavenArtifactController` by separating reused code in `GET` and `HEAD`.

### CODE REVIEW

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/GroupRepositoryProvider.java [hunk 1]
```diff
@@ -196,6 +197,109 @@ private ArtifactInputStream getInputStream(Repository repository, String artifac
         return is;
     }
 
+    public RepositoryPath getPath(String storageId,
+                                       String repositoryId,
+                                       String path)
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/GroupRepositoryProvider.java]: Cosmetic: The parameters above aren't indented properly.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/GroupRepositoryProvider.java [hunk 2]
```diff
@@ -196,6 +197,109 @@ private ArtifactInputStream getInputStream(Repository repository, String artifac
...
+   {
+        final Storage storage = getConfiguration().getStorage(storageId);
+        final Repository groupRepository = storage.getRepository(repositoryId);
+        final LayoutProvider layoutProvider = layoutProviderRegistry.getProvider(groupRepository.getLayout());
+        RepositoryPath artifactPath = null;
+        
+        artifactPath = layoutProvider.resolve(groupRepository).resolve(path);
+        
+        if(Files.exists(artifactPath))
+                return artifactPath;
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/GroupRepositoryProvider.java]: * Indentation
* Please, always use curly braces for such statements

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/GroupRepositoryProvider.java [hunk 3]
```diff
@@ -196,6 +197,109 @@ private ArtifactInputStream getInputStream(Repository repository, String artifac
...
+        return null;
+    }
+    
+    private RepositoryPath getPath(Repository repository, String path)
+            throws IOException,
+                   NoSuchAlgorithmException,
+                   ArtifactTransportException,
+                   ProviderImplementationException
+    {
+        RepositoryProvider provider = getRepositoryProviderRegistry().getProvider(repository.getType());
```
[the remark this answers is not in the document]
Comment (carlspring) [reply strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/GroupRepositoryProvider.java]: If you won't be using the commented out code below, then Sonar is right about this.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/ArtifactManagementService.java [hunk 4]
```diff
@@ -67,5 +68,9 @@ void removeTimestampedSnapshots(String storageId,
             throws IOException;
 
     Storage getStorage(String storageId);
+
+    RepositoryFileAttributes getAttributes(String storageId,
+                                           String repositoryId,
+                                           String path) throws ArtifactTransportException, ProviderImplementationException, NoSuchAlgorithmException;
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/ArtifactManagementService.java]: Cosmetic: Please, split the `throws` clause on a separate line.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/ArtifactResolutionService.java [hunk 5]
```diff
@@ -42,4 +43,8 @@ URL resolveArtifactResource(String storageId,
             throws MalformedURLException, 
                    IOException;
 
+    RepositoryFileAttributes getAttributes(String storageId,
+                                           String repositoryId,
+                                           String path) throws IOException, NoSuchAlgorithmException, ArtifactTransportException, ProviderImplementationException;
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/ArtifactResolutionService.java]: Cosmetic: Please, split the `throws` list clause on separate lines.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/AbstractArtifactManagementService.java [hunk 6]
```diff
@@ -419,4 +420,32 @@ public void delete(String storageId,
         }
     }
 
+    @Override
+    public RepositoryFileAttributes getAttributes(String storageId,
+                               String repositoryId,
+                               String path)
+            throws ArtifactTransportException,
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/AbstractArtifactManagementService.java]: Cosmetic: Please, split the `throws` list clause on separate lines.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/AbstractArtifactManagementService.java [hunk 7]
```diff
@@ -419,4 +420,32 @@ public void delete(String storageId,
         }
     }
 
+    @Override
+    public RepositoryFileAttributes getAttributes(String storageId,
+                               String repositoryId,
+                               String path)
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/AbstractArtifactManagementService.java]: Cosmetic: The parameters above aren't indented properly.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/AbstractArtifactManagementService.java [hunk 8]
```diff
@@ -419,4 +420,32 @@ public void delete(String storageId,
...
+        {
+            try
+            {
+                return artifactResolutionService.getAttributes(storageId, repositoryId, path);
+            }
+            catch (IOException e)
+            {
+                // This is not necessarily an error. It could simply be a check
+                // whether a resource exists, before uploading/updating it.
+                logger.debug("The REQUESTED path does not exist: - /" + storageId + "/" + repositoryId + "/" + path);
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/AbstractArtifactManagementService.java]: `logger.debug("The requested path does not exist: /" + storageId + "/" + repositoryId + "/" + path);`

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/AbstractArtifactManagementService.java [hunk 9]
```diff
@@ -419,4 +420,32 @@ public void delete(String storageId,
...
+            }
+        }
+        catch (Exception e)
+        {
+            // TODO Auto-generated catch block
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/AbstractArtifactManagementService.java]: Please, use `logger.error` for this (and remove the `// TODO` line).

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ArtifactResolutionServiceImpl.java [hunk 10]
```diff
@@ -130,5 +138,49 @@ public URL resolveArtifactResource(String storageId,
...
+        RepositoryProvider repositoryProvider = repositoryProviderRegistry.getProvider(repository.getType());
+        
+        logger.debug("REpository TYpe = "+repositoryProvider.getAlias());
+        
+        RepositoryPath path = null;
+        
+        try
+        {
+            if(!repositoryProvider.getAlias().equals("group"))
+                path = layoutProvider.resolve(repository).resolve(artifactPath);
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ArtifactResolutionServiceImpl.java]: Please always use curly braces. This makes sure that there won't be misunderstandings in condition checking (as this *can* happen).

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ArtifactResolutionServiceImpl.java [hunk 11]
```diff
@@ -130,5 +138,49 @@ public URL resolveArtifactResource(String storageId,
...
+            else
+                path = ((GroupRepositoryProvider)repositoryProvider).getPath(storageId,repositoryId,artifactPath);
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ArtifactResolutionServiceImpl.java]: * Same as above.
* Also, please always use spaces after `+` and `,`, as this is the code style of the project. Thanks! :)

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ArtifactResolutionServiceImpl.java [hunk 12]
```diff
@@ -130,5 +138,49 @@ public URL resolveArtifactResource(String storageId,
...
+        }
+        catch (Exception e)
+        {
+            // TODO Auto-generated catch block
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ArtifactResolutionServiceImpl.java]: Same as my earlier comment regarding `logger.error`.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ArtifactResolutionServiceImpl.java [hunk 13]
```diff
@@ -130,5 +138,49 @@ public URL resolveArtifactResource(String storageId,
...
+        logger.debug("******");
+        
+        artifactOperationsValidator.validate(storageId, repositoryId, artifactPath);
+
+        final Repository repository = getStorage(storageId).getRepository(repositoryId);
+     
+        LayoutProvider<?> layoutProvider = layoutProviderRegistry.getProvider(repository.getLayout());
+        RepositoryProvider repositoryProvider = repositoryProviderRegistry.getProvider(repository.getType());
+        
+        logger.debug("REpository TYpe = "+repositoryProvider.getAlias());
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ArtifactResolutionServiceImpl.java]: Cosmetic: Please, change this to:
```
logger.debug("Repository type = " + repositoryProvider.getAlias());
```

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java [hunk 14]
```diff
@@ -200,14 +202,173 @@ public void download(@ApiParam(value = "The storageId", required = true)
...
+    @PreAuthorize("hasAuthority('ARTIFACTS_RESOLVE')")
+    @RequestMapping(value = { "{storageId}/{repositoryId}/{path:.+}" }, method = RequestMethod.HEAD)
+    public void getHeaders(@ApiParam(value = "The storageId", required = true)
+                         @PathVariable String storageId,
+                         @ApiParam(value = "The repositoryId", required = true)
+                         @PathVariable String repositoryId,
+                         @RequestHeader HttpHeaders httpHeaders,
+                         @PathVariable String path,
+                         HttpServletRequest request,
+                         HttpServletResponse response)
```
Comment (carlspring) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java]: Cosmetic: Please, re-indent the parameters.

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java [hunk 15]
```diff
@@ -200,14 +202,173 @@ public void download(@ApiParam(value = "The storageId", required = true)
...
+            }
+
+            if (isRangedRequest(httpHeaders))
+            {
+                logger.debug("Detecting range request....");
+
+                handlePartialDownload(is, httpHeaders, response);
+            }
+
+            artifactEventListenerRegistry.dispatchArtifactDownloadingEvent(storage.getId(), repository.getId(), path);
```
Comment (carlspring) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java]: The flow here with the dispatching of these two events doesn't look right.

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java [hunk 16]
```diff
@@ -200,14 +202,173 @@ public void download(@ApiParam(value = "The storageId", required = true)
...
+
+        int readLength;
+        byte[] bytes = new byte[4096];
+        while ((readLength = ais.read(bytes, 0, bytes.length)) != -1)
+        {
+            totalBytes += readLength;
+        }
+        response.setHeader("Content-Length", Long.toString(totalBytes));
+        
+        response.setHeader("Last-Updated",fileAttributes.lastModifiedTime().toString());
```
Comment (carlspring) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java]: Cosmetic: Spaces between parameters, please.

File: strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/maven/MavenArtifactControllerTest.java [hunk 17]
```diff
@@ -279,7 +282,40 @@ private void resolveArtifact(String artifactPath,
...
         assertEquals("SHA-1 checksums did not match!", sha1Local, sha1Remote);
     }
-
+    
+    @Test
+    public void testHeaderFetch()
+            throws Exception
+    {   
+        String artifactPath1 = "storages/storage0/act-releases-1/org/carlspring/strongbox/browse/foo-bar/2.4/foo-bar-2.4.jar";
+
```
Comment (carlspring) [inline strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/maven/MavenArtifactControllerTest.java]: Cosmetic: No need for so many empty lines (all over the place).

File: strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/maven/MavenArtifactControllerTest.java [hunk 18]
```diff
@@ -279,7 +282,40 @@ private void resolveArtifact(String artifactPath,
...
+        
+       
+       
+        Headers headers1 = client.getHeaders(artifactPath1);
+       
+        
+        
+        String artifactPath2 = "storages/storage0/act-releases-1/org/carlspring/strongbox/browse/foo-bar/2.4/foo-bar-2.4.pom";
+        
+        logger.debug("Pausing now ");
```
Comment (carlspring) [inline strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/maven/MavenArtifactControllerTest.java]: Is there anything actually sleeping here? (Like, for example, `Thread.sleep(1000)`)...?

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java (hunk 15 above)
Comment (sanket407) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java]: I do not get what you mean by 'flow'.
It's the same steps as in GET only removed the copying of artifact to Response Body

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java [hunk 19]
```diff
@@ -200,14 +202,177 @@ public void download(@ApiParam(value = "The storageId", required = true)
...
+            throws Exception
+    {
+        logger.debug("Requested Headers for /" + storageId + "/" + repositoryId + "/" + path + ".");
+
+        Storage storage = configurationManager.getConfiguration().getStorage(storageId);
+        if (storage == null)
+        {
+            logger.error("Unable to find storage by ID " + storageId);
+
+            response.sendError(INTERNAL_SERVER_ERROR.value(), "Unable to find storage by ID " + storageId);
```
Comment (carlspring) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java]: This should be  a `404` error.
 You should not be returning error `500`, as it's more for `RuntimeException`-s and the likes.

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java (hunk 15 above)
Comment (carlspring) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java]: Oh, so the code for this download has now been extracted inside a listener?

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/nuget/NugetPackageController.java [hunk 20]
```diff
@@ -388,7 +391,147 @@ private void getPackageInternal(String storageId,
...
+                                       packageVersion),
+                                       e);
+
+            response.setStatus(INTERNAL_SERVER_ERROR.value());
+            return;
+        }
+        
+        setHeaders(is, response, storageId, repositoryId, path, fileName);
+
+        logger.debug("Header Download succeeded.");
```
Comment (carlspring) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/nuget/NugetPackageController.java]: This logging message could be removed.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/GroupRepositoryProvider.java [hunk 21]
```diff
@@ -192,10 +194,107 @@ private ArtifactInputStream getInputStream(Repository repository, String artifac
         ArtifactInputStream is = provider.getInputStream(repository.getStorage().getId(),
                                                          repository.getId(),
                                                          artifactPath);
-
         return is;
     }
 
+    public RepositoryPath getPath(String storageId,
+                                       String repositoryId,
+                                       String artifactPath)
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/GroupRepositoryProvider.java]: Cosmetic: Please, re-indent the parameters.

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java [hunk 22]
```diff
@@ -200,14 +202,177 @@ public void download(@ApiParam(value = "The storageId", required = true)
...
+
+            return;
+        }
+
+        Repository repository = storage.getRepository(repositoryId);
+        if (repository == null)
+        {
+            logger.error("Unable to find repository by ID " + repositoryId + " for storage " + storageId);
+
+            response.sendError(INTERNAL_SERVER_ERROR.value(),
```
Comment (carlspring) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java]: Same as above:

This should be  a `404` error.
You should not be returning error `500`, as it's more for `RuntimeException`-s and the likes.

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java (hunk 19 above)
Comment (sanket407) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java]: I think we have another issue which deals with this specifically across all rest controllers !
#418

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/nuget/NugetPackageController.java [hunk 23]
```diff
@@ -388,7 +391,147 @@ private void getPackageInternal(String storageId,
             response.setStatus(INTERNAL_SERVER_ERROR.value());
         }
     }
+    
+    @ApiOperation(value = "Used to download the headers for a package")
+    @ApiResponses(value = { @ApiResponse(code = HttpURLConnection.HTTP_OK, message = "The package was downloaded successfully."),
+                            @ApiResponse(code = HttpURLConnection.HTTP_INTERNAL_ERROR, message = "An error occurred.") })
+    @PreAuthorize("hasAuthority('ARTIFACTS_RESOLVE')")
+    @RequestMapping(path = "{storageId}/{repositoryId}/{commandName:(?:download|package)}/{packageId}/{packageVersion}", method = RequestMethod.HEAD, produces = MediaType.APPLICATION_OCTET_STREAM)
+    public void downloadPackageHeaders(@ApiParam(value = "The storageId", required = true) @PathVariable(name = "storageId") String storageId,
```
Comment (carlspring) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/nuget/NugetPackageController.java]: This method should be called `getPackageHeaders`, as you're not downloading headers, you're getting them.

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/nuget/NugetPackageController.java [hunk 24]
```diff
@@ -388,7 +391,147 @@ private void getPackageInternal(String storageId,
...
+    {
+        getHeaders(storageId, repositoryId, packageId, packageVersion, response);
+    }
+    
+    @ApiOperation(value = "Used to download the headers for a package")
+    @ApiResponses(value = { @ApiResponse(code = HttpURLConnection.HTTP_OK, message = "The package was downloaded successfully."),
+                            @ApiResponse(code = HttpURLConnection.HTTP_INTERNAL_ERROR, message = "An error occurred.") })
+    @PreAuthorize("hasAuthority('ARTIFACTS_RESOLVE')")
+    @RequestMapping(path = "{storageId}/{repositoryId}/{packageId}/{packageVersion}", method = RequestMethod.HEAD, produces = MediaType.APPLICATION_OCTET_STREAM)
+    public void getPackageHeaders(@ApiParam(value = "The storageId", required = true) @PathVariable(name = "storageId") String storageId,
```
Comment (carlspring) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/nuget/NugetPackageController.java]: What's the difference between this method and `downloadPackageHeaders`?

Comment (sanket407) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/nuget/NugetPackageController.java]: Path

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java (hunk 19 above)
Comment (carlspring) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java]: That issue is different. It's not about non-existing storages, or repositories, it's about paths underneath them.

File: strongbox-rest-client/src/test/java/org/carlspring/strongbox/rest/client/RestAssuredArtifactClient.java [hunk 25]
```diff
@@ -366,4 +366,58 @@ public String search(String repositoryId,
                            .asString();
     }
 
+    public Headers getHeadersfromHEAD(String url)
```
Comment (sbespalov) [inline strongbox-rest-client/src/test/java/org/carlspring/strongbox/rest/client/RestAssuredArtifactClient.java]: I think this should be `getHeadersFromHEAD`

Comment (carlspring) [reply strongbox-rest-client/src/test/java/org/carlspring/strongbox/rest/client/RestAssuredArtifactClient.java]: +1

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/GroupRepositoryProvider.java [hunk 26]
```diff
@@ -67,74 +68,38 @@ public String getAlias()
...
+        logger.debug("Resolved path for Group Repository = " + artifactPath);
+        
+        Repository repository = artifactPath.getFileSystem().getRepository();
+        String resolvedStorageId = repository.getStorage().getId();
+        String resolvedRepositoryId = repository.getId();
+        String resolvedPath = artifactPath.relativize().toString();
+        RepositoryProvider provider = getRepositoryProviderRegistry().getProvider(repository.getType());
+        
+        //If current repository is a group repository present in local cache
+        if(getAlias().equals(provider.getAlias()) && Files.exists(artifactPath))
```
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/GroupRepositoryProvider.java]: we can't have artifact files physically stored in group repository folder, so this condition will never be true

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/GroupRepositoryProvider.java [hunk 27]
```diff
@@ -270,11 +287,11 @@ public ArtifactOutputStream getOutputStream(String storageId,
 
                 // count coordinates intersection
                 groupLimit += repositoryResult.stream()
-                                              .map((p) -> resultMap.put(getArtifactCoordinates(p),
-                                                                        p))
-                                              .filter(p ->  p != null)
-                                              .collect(Collectors.toList())
-                                              .size();
+                        .map((p) -> resultMap.put(getArtifactCoordinates(p),
```
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/GroupRepositoryProvider.java]: indentation is incorrect

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/GroupRepositoryProvider.java [hunk 28]
```diff
@@ -308,11 +325,11 @@ public ArtifactOutputStream getOutputStream(String storageId,
                                                       boolean traverse)
     {
         Set<Repository> result = groupRepository.getGroupRepositories()
-                                                .stream()
-                                                .map(groupRepoId -> {
-                                                    return getRepository(groupRepository.getStorage(), groupRepoId);
-                                                })
-                                                .collect(Collectors.toCollection(LinkedHashSet::new));
+                .stream()
```
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/GroupRepositoryProvider.java]: same as above

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/GroupRepositoryProvider.java [hunk 29]
```diff
@@ -368,12 +385,14 @@ public Long count(RepositorySearchRequest searchRequest)
 
         Set<Pair<String, String>> storageRepositoryPairSet = collectGroupRepositorySet(groupRepository,
                                                                                        true).stream()
-                                                                                            .map(r -> Pair.with(r.getStorage()
-                                                                                                                 .getId(),
-                                                                                                                r.getId()))
-                                                                                            .collect(Collectors.toCollection(LinkedHashSet::new));
+                .map(r -> Pair.with(r.getStorage()
```
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/GroupRepositoryProvider.java]: here too

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/HostedRepositoryProvider.java [hunk 30]
```diff
@@ -55,24 +57,28 @@ public String getAlias()
     public ArtifactInputStream getInputStream(String storageId,
```
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/HostedRepositoryProvider.java]: I think that this method implementation should use `getPath` as we agreed in channel

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/HostedRepositoryProvider.java [hunk 31]
```diff
@@ -55,24 +57,28 @@ public String getAlias()
...
+            return (ArtifactInputStream)Files.newInputStream(artifactPath);
         }
 
         logger.debug("The artifact {} was not found in the local cache", artifactPath);
         return null;
     }
-
+        
     @Override
     public ArtifactOutputStream getOutputStream(String storageId,
```
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/HostedRepositoryProvider.java]: this should use `getPath` too

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/proxied/LocalStorageProxyRepositoryArtifactResolver.java [hunk 32]
```diff
@@ -61,7 +62,10 @@
     @Override
     protected InputStream preProxyRepositoryAccessAttempt(final Repository repository,
                                                           final String path)
-            throws IOException
```
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/proxied/LocalStorageProxyRepositoryArtifactResolver.java]: why this changes were needed?

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/proxied/ProxyRepositoryArtifactResolver.java [hunk 33]
```diff
@@ -121,7 +120,10 @@ protected InputStream onSuccessfulProxyRepositoryResponse(final InputStream is,
 
     protected InputStream preProxyRepositoryAccessAttempt(Repository repository,
                                                           String path)
-            throws IOException
+            throws IOException,
```
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/proxied/ProxyRepositoryArtifactResolver.java]: same as above

File: strongbox-cron/strongbox-cron-rest/src/test/java/org/carlspring/strongbox/cron/controller/CronTaskConfigurationControllerTest.java [hunk 34]
```diff
@@ -30,6 +31,7 @@
  */
 @CronTaskRestTest
 @RunWith(SpringJUnit4ClassRunner.class)
+@Ignore
```
Comment (fuss86) [inline strongbox-cron/strongbox-cron-rest/src/test/java/org/carlspring/strongbox/cron/controller/CronTaskConfigurationControllerTest.java]: `@Ignore`

Could you please verify this test once again without ignoring it ?

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/GroupRepositoryProvider.java [hunk 35]
```diff
@@ -67,74 +68,38 @@ public String getAlias()
...
+        String resolvedPath = artifactPath.relativize().toString();
+        RepositoryProvider provider = getRepositoryProviderRegistry().getProvider(repository.getType());
+        
+        //If current repository is a group repository present in local cache
+        if(getAlias().equals(provider.getAlias()) && Files.exists(artifactPath))
         {
-            return null;
+            return (ArtifactInputStream)Files.newInputStream(artifactPath);
         }
-        return artifactFileAttributes.isMetadata() ?
```
Comment (fuss86) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/GroupRepositoryProvider.java]: `artifactFileAttributes.isMetadata()`

I'm worried that the logic checking/resolving *only* metadata files directly from group repository file system path is gone in your Pull Request.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/HostedRepositoryProvider.java [hunk 36]
```diff
@@ -129,4 +137,25 @@ public Long count(RepositorySearchRequest searchRequest)
...
+                                  String path)
+                 throws IOException,
+                        NoSuchAlgorithmException,
+                        ArtifactTransportException,
+                        ProviderImplementationException
+    {
+        Repository repository = getConfiguration().getStorage(storageId).getRepository(repositoryId);
+
+        final LayoutProvider layoutProvider = layoutProviderRegistry.getProvider(repository.getLayout());
+        final RepositoryPath artifactPath = layoutProvider.resolve(repository).resolve(path);
```
Comment (fuss86) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/HostedRepositoryProvider.java]: You can move this line inside the `if` body below, when `if` condition is `true`

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/HostedRepositoryProvider.java [hunk 37]
```diff
@@ -129,4 +137,25 @@ public Long count(RepositorySearchRequest searchRequest)
...
     }
 
+    @Override
+    public RepositoryPath getPath(String storageId,
+                                  String repositoryId,
+                                  String path)
+                 throws IOException,
+                        NoSuchAlgorithmException,
+                        ArtifactTransportException,
+                        ProviderImplementationException
```
Comment (fuss86) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/HostedRepositoryProvider.java]: `throws IOException,
NoSuchAlgorithmException,
ArtifactTransportException,
ProviderImplementationException
`

I doubt all of them are necessary here :)

File: strongbox-storage/strongbox-storage-core/src/main/java/org/carlspring/strongbox/providers/repository/RepositoryProvider.java [hunk 38]
```diff
@@ -34,5 +37,13 @@ ArtifactOutputStream getOutputStream(String storageId,
...
     
     Long count(RepositorySearchRequest searchRequest);
+    
+    Path getPath(String storageId,
+                 String repositoryId,
+                 String artifactPath)
+           throws IOException,
+                  NoSuchAlgorithmException,
+                  ArtifactTransportException,
+                  ProviderImplementationException;
```
Comment (fuss86) [inline strongbox-storage/strongbox-storage-core/src/main/java/org/carlspring/strongbox/providers/repository/RepositoryProvider.java]: `
throws IOException,
NoSuchAlgorithmException,
ArtifactTransportException,
ProviderImplementationException;
`
I strongly believe it's too much just for getting path from repository provider.

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java [hunk 39]
```diff
@@ -164,8 +238,8 @@ public void download(@ApiParam(value = "The storageId", required = true)
 
                 response.setStatus(INTERNAL_SERVER_ERROR.value());
             }
-
-            return;
+            logger.debug("xxx");
```
Comment (fuss86) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java]: `logger.debug("xxx");`

Are you a big fan of Vin Diesel ? :>

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java [hunk 40]
```diff
@@ -197,16 +269,70 @@ public void download(@ApiParam(value = "The storageId", required = true)
...
+            long totalBytes = 0L;
+            int readLength;
+            byte[] bytes = new byte[4096];
+            
+            while ((readLength = ais.read(bytes, 0, bytes.length)) != -1)
+            {
+                totalBytes += readLength;
+            }
+        
+            response.setHeader("Content-Length", Long.toString(totalBytes));
```
Comment (fuss86) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java]: @carlspring I don't think counting input stream bytes just for displaying result as a `Content-Length` header is a good approach here.

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java [hunk 41]
```diff
@@ -119,6 +123,76 @@ public void download(@ApiParam(value = "The storageId", required = true)
...
+                                                       response,
+                                                       request,
+                                                       httpHeaders);
+       
+       if(is == null)
+       {
+           return;
+       }
+       
+       RepositoryFileAttributes fileAttributes = getRepositoryFileAttributes(storageId,
```
Comment (fuss86) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java]: `getRepositoryFileAttributes` may return `null` as far as I see. What then ? You don't handle it, or I am missing something ?

Comment (carlspring) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java]: I had not better idea about guessing the size. I mean... once you're done writing to the stream, you'll have the final number of bytes.

Do you know of a better way?

Comment (fuss86) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java]: We have 

`org.carlspring.strongbox.domain.ArtifactEntry#sizeInBytes` or `org.carlspring.strongbox.providers.io.RepositoryFileAttributes#size`

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java (hunk 40 above)
Comment (carlspring) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java]: I had not better idea about guessing the size. I mean... once you're done writing to the stream, you'll have the final number of bytes.

Do you know of a better way?

Comment (fuss86) [reply strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java]: We have 

`org.carlspring.strongbox.domain.ArtifactEntry#sizeInBytes` or `org.carlspring.strongbox.providers.io.RepositoryFileAttributes#size`

**Comment by carlspring** [comment]
Will be covered in #491.
